File: src/0904_fruit_into_baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {
    let count = new Map();
    let [left, total, res] = [0, 0, 0];

    for (fruit of fruits) {
        count.set(fruit, (count.get(fruit) || 0) + 1);
        total++;

        while (count.size > 2) {
            let f = fruits[left];
            count.set(f, count.get(f) - 1);
            total -= 1;
            left += 1;
            if (!count.get(f)) {
                count.delete(f);
            }
        }
        res = Math.max(res, total);
    }

    return res;
};

module.exports = { totalFruit };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(totalFruit([1, 2, 1]), 3);
    assert.deepStrictEqual(totalFruit([0, 1, 2, 2]), 3);
    assert.deepStrictEqual(totalFruit([1, 2, 3, 2, 2]), 4);
}

File: src/0904_fruit_into_baskets.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn total_fruit(fruits: Vec<i32>) -> i32 {
        let mut count = HashMap::new();
        let (mut left, mut total, mut res) = (0, 0, 0);

        for fruit in &fruits {
            *count.entry(*fruit).or_insert(0) += 1;
            total += 1;

            while count.len() > 2 {
                let f = fruits[left];
                match count.remove(&f) {
                    Some(v) if v > 1 => {
                        count.insert(f, v - 1);
                    }
                    _ => {}
                }
                total -= 1;
                left += 1;
            }

            res = res.max(total);
        }

        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn main_asserts() {
        assert_eq!(Solution::total_fruit(vec![1, 2, 1]), 3);
        assert_eq!(Solution::total_fruit(vec![0, 1, 2, 2]), 3);
        assert_eq!(Solution::total_fruit(vec![1, 2, 3, 2, 2]), 4);
    }
}
